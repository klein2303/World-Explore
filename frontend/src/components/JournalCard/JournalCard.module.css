/* Card Wrapper Styles */
.cardWrapper {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin-bottom: 30px;
}

/* Card Styles */
.card {
    position: relative;
    width: 150px;
    height: 200px;
    background-color: #e4e1f7;
    border-radius: 3px 8px 8px 3px;
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
    overflow: hidden;
    cursor: pointer;
}

body[data-theme="dark"] .card {
    background-color: #6b6787;
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.3);
}

.verticalStrip {
    position: absolute;
    top: 0;
    width: 8px;
    height: 100%;
    background-color: #424242;
    z-index: 1;
}

body[data-theme="dark"] .verticalStrip {
    background-color: #bebebe;
}

.cardImage {
    width: 100%;
    height: 100%;
    object-fit: cover;
    object-position: center;
}

.cardText {
    text-align: center;
    margin-top: 12px;
}

.cardTitle {
    font-size: 18px;
    font-weight: 400;
    color: #424242;
}

body[data-theme="dark"] .cardTitle {
    color: white;
}

.cardDate {
    font-size: 13px;
    color: gray;
}

/* Overlay Text Styles */
.overlayText {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    color: white;
    font-size: 16px;
    font-weight: bold;
    text-align: center;
    background-color: rgba(0, 0, 0, 0.5);
    border-radius: 6px;
    padding: 8px;
    opacity: 0;
}

.card:hover .overlayText {
    opacity: 1;
}

/* Modal Styles */
.modal {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.5);
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 3;
}

.modalContent {
    background-color: white;
    padding: 20px;
    border-radius: 10px;
    width: 500px;
    max-width: 90%;
}

.textArea {
    width: 100%;
    height: 200px;
    margin-top: 10px;
    padding: 10px;
    border: 1px solid #ccc;
    border-radius: 5px;
    box-sizing: border-box;
}

.closeButton {
    background-color: red;
    color: white;
    padding: 8px 16px;
    border: none;
    border-radius: 6px;
    cursor: pointer;
    margin-top: 10px;
}
