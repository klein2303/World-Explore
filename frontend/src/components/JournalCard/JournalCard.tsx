import {useState} from "react";
import styles from "./JournalCard.module.css";

interface JournalCardProps {
    country: string;
    date: string | null; // Null if visited, but not journaled
    image: string;
}

const JournalCard = ({country, date, image}: JournalCardProps) => {
    const [isWriting, setIsWriting] = useState<boolean>(false); // To handle modal visibility

    // Function to toggle the modal
    const toggleModal = () => {
        if (!date) {
            // Only toggle the modal if there is no date (not journaled)
            setIsWriting(!isWriting);
        }
    };

    return (
        <article className={styles.cardWrapper} aria-label={`Journal card wrapper for ${country}`}>
            <section className={styles.card} onClick={toggleModal} aria-label={`Journal card for ${country}`}>
                <div className={styles.verticalStrip} role="presentation"></div>
                <img src={image} alt={`Image of ${country}`} className={styles.cardImage} />

                {/* Overlay text */}
                <div className={styles.overlayText} role="text">{date ? "Read Journal" : "Write Journal"}</div>
            </section>
            <div className={styles.cardText} role="contentinfo" aria-label={`Journal card text for ${country}`}>
                <header className={styles.cardTitle}>{country}</header>

                {/* Display the date if the country has a journal entry */}
                {date && <p className={styles.cardDate}>{date}</p>}

                {/* Modal for writing a journal */}
                {isWriting && (
                    <aside className={styles.modal} role="dialog" aria-modal="true" aria-labelledby="journalModalTitle">
                        <div className={styles.modalContent} role="document">
                            <h2 id="journalModalTitle">Write your journal entry for {country}</h2>
                            <textarea
                                placeholder="Start writing..."
                                className={styles.textArea}
                                aria-label="Journal entry text area"></textarea>
                            <button onClick={toggleModal} className={styles.closeButton} aria-label="Close modal">
                                Close
                            </button>
                        </div>
                    </aside>
                )}
            </div>
        </article>
    );
};

export default JournalCard;
