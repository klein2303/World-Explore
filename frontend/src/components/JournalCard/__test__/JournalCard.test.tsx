import {render, screen, fireEvent} from "@testing-library/react";
import {expect, it, describe} from "vitest";
import JournalCard from "../JournalCard";

describe("JournalCard", () => {
    const mockProps = {
        country: "France",
        date: "2023-01-12",
        image: "https://example.com/france.jpg",
    };

    const mockPropsUnwritten = {
        country: "Spain",
        date: null,
        image: "https://example.com/spain.jpg",
    };

    it("renders correctly with a journal entry", () => {
        const {asFragment} = render(<JournalCard {...mockProps} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly without a journal entry", () => {
        const {asFragment} = render(<JournalCard {...mockPropsUnwritten} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("displays the correct overlay text based on the date", () => {
        render(<JournalCard {...mockProps} />);
        expect(screen.getByText("Read Journal")).toBeInTheDocument();

        render(<JournalCard {...mockPropsUnwritten} />);
        expect(screen.getByText("Write Journal")).toBeInTheDocument();
    });

    it("toggles the modal visibility when clicked", () => {
        render(<JournalCard {...mockPropsUnwritten} />);
        const card = screen.getByLabelText("Journal card for Spain");
        fireEvent.click(card);
        expect(screen.getByRole("dialog")).toBeInTheDocument();

        const closeButton = screen.getByLabelText("Close modal");
        fireEvent.click(closeButton);
        expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
    });

    it("does not toggle the modal visibility when there is a date", () => {
        render(<JournalCard {...mockProps} />);
        const card = screen.getByLabelText("Journal card for France");
        fireEvent.click(card);
        expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
    });
});
