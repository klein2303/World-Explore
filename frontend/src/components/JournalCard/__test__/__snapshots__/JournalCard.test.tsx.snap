// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JournalCard > renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <article
    aria-label="Journal card wrapper for Japan"
    class="_cardWrapper_be814f"
  >
    <a
      href="/JournalPage/Japan"
    >
      <section
        aria-label="Journal card for Japan"
        class="_card_be814f"
      >
        <div
          class="_verticalStrip_be814f"
          role="presentation"
        />
        <img
          alt="Image of Japan"
          class="_cardImage_be814f"
          src="https://example.com/japan.jpg"
        />
        <div
          class="_overlayText_be814f"
          role="text"
        >
          Write Journal
        </div>
      </section>
    </a>
    <div
      aria-label="Journal card text for Japan"
      class="_cardText_be814f"
      role="contentinfo"
    >
      <header
        class="_cardTitle_be814f"
      >
        Japan
      </header>
    </div>
  </article>
</DocumentFragment>
`;
